### ISSUE TITLE
__class__.__add__() confuses type detection?

### ISSUE THREAD

**Issue by karolyi** [microsoft/pyright#3848]
**Describe the bug**
Hey, sorry to keep bugging you but I think I've found another issue where pyright wrongly determines the type of a return variable. I've found this also with `django-money` where upon an addition the type would mysteriously change and as such, pyright would complain about type incompatibilty. The gist of it is, when using a class with an `__add__` function, pyright can't determine the exact return type by itself, whereas when using about the same functionality in a normal function, it can.

**To Reproduce**
I've prepared a little code to demonstrate:
```python
from __future__ import annotations

from datetime import date, timedelta


class RootClass(object):

    def __init__(self, val: int):
        self._val = val

    def __add__(self, other) -> RootClass:
        return self.__class__(val=self._val + other._val)


class TestClass(RootClass):
    def __add__(self, other):
        if isinstance(other, date):
            x = other.__add__(timedelta(days=self._val))
            return x
        result = super().__add__(other)
        return result

    def __str__(self):
        return f'{self.__class__}: {self._val}'


def working_type(var):
    if isinstance(var, date):
        x = var.__add__(timedelta(days=1))
        return x
    result = var + 1
    return result


x = TestClass(val=5)
bad_1 = x + date(2022, 6, 1)
bad_2 = x + TestClass(val=1)
print(type(bad_1), type(bad_2))

good_1 = working_type(5)
good_2 = working_type(date(2022, 6, 1))
print(good_1, good_2)
```

**Expected behavior**
The class with the `__add__`, its return type should be recognized right just like with the non-class example.

**Screenshots or Code**
See the screenshots with pointing on the variables:
![image](https://user-images.githubusercontent.com/987055/185743557-e430457d-3d49-4f2d-bf81-801fc6fe32f4.png)
![image](https://user-images.githubusercontent.com/987055/185743572-a535611d-d1f9-4513-b857-9d9ba79a6596.png)
![image](https://user-images.githubusercontent.com/987055/185743592-21cc7aaf-8852-49b1-8f8f-eab306f91d5a.png)
![image](https://user-images.githubusercontent.com/987055/185743608-a29760c9-dd24-4708-8773-9c7e03727203.png)

**Comment by erictraut** [comment]
This is working as expected.

The functions `working_type` and `__add__` are both unannotated. In particular, the input parameter `other` has no type annotation. When pyright encounters such a function, it uses a variety of techniques to attempt to infer the return type of a call that targets that function. These techniques vary depending on how the function is invoked — whether it's a direct call (as in the case of `working_type`) or an implicit invocation through a magic method like `__add__` or `__radd__`. In the case of the explicit call, pyright uses an advanced inference technique called [call-site return type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#call-site-return-type-inference).

If you want this to type check correctly, you should provide appropriate type annotations for input parameters. And since `__add__` and `working_type` are polymorphic functions whose return types depend on the input types, you will also need to include `@overload` signatures as specified in PEP 484.
